metatheory/TestNEAT.hs

```diff
@@ -0,0 +1,105 @@
+import           MAlonzo.Code.Scoped                          (deBruijnifyK, unDeBruijnifyK)
+
+import           Language.PlutusCore.DeBruijn
+import           Raw
+
+
+main :: IO ()
+main = defaultMain allTests
+
+allTests :: TestTree
+allTests = testGroup "all tests"
+  [ testCaseCount "soundness" $
+      testTyProp depth kind prop_checkKindSound
+  , testCaseCount "normalization" $
+      testTyProp depth kind prop_normalizePreservesKind
+  , testCaseCount "normalizationSound" $
+      testTyProp depth kind prop_normalizeTypeSound
+  , testCaseCount "normalizationAgree" $
+      testTyProp depth kind prop_normalizeTypeSame
+  , testCaseCount "kindInferAgree" $
+      testTyProp depth kind prop_kindInferSame
+  ]
+
+testCaseCount :: String -> IO Integer -> TestTree
+testCaseCount name act = testCaseInfo name $
+  liftM (\i -> show i ++ " types generated") act
+
+
+-- NEAT settings
+
+depth :: Int
+depth = 10
+
+kind :: Kind ()
+kind = Type ()
+
+-- |Check if the type/kind checker or generation threw any errors.
+isSafe :: ExceptT (ErrorP a) Quote a -> Cool
+isSafe = toCool . isRight . runQuote . runExceptT
+
+prop_checkKindSound :: TyProp
+prop_checkKindSound k _ tyQ = isSafe $ do
+  ty <- withExceptT GenErrorP tyQ
+  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
+  withExceptT TypeErrorP $ case checkKindAgda (AlexPn 0 0 0 <$ tyDB) (deBruijnifyK (convK k)) of
+    Just _  -> return ()
+    Nothing -> throwError undefined -- TODO
+
+prop_normalizePreservesKind :: TyProp
+prop_normalizePreservesKind k _ tyQ = isSafe $ do
+  ty  <- withExceptT GenErrorP tyQ
+  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
+  tyN <- withExceptT TypeErrorP $ case normalizeTypeAgda (AlexPn 0 0 0 <$ tyDB) of
+    Just tyN -> return tyN
+    Nothing  -> throwError undefined -- TODO
+  withExceptT TypeErrorP $ case checkKindAgda (AlexPn 0 0 0 <$ tyN) (deBruijnifyK (convK k)) of
+    Just _  -> return ()
+    Nothing -> throwError undefined -- TODO
+
+-- the agda implementation throws names away, so I guess we need to compare deBruijn terms
+prop_normalizeTypeSound :: TyProp
+prop_normalizeTypeSound k tyG tyQ = isSafe $ do
+  ty <- withExceptT GenErrorP tyQ
+  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
+  tyN1 <- withExceptT TypeErrorP $ case normalizeTypeAgda (AlexPn 0 0 0 <$ tyDB) of
+    Just tyN -> return tyN
+    Nothing  -> throwError undefined -- TODO
+  ty1 <- withExceptT FVErrorP $ unDeBruijnTy tyN1
+
+  ty2 <- withExceptT GenErrorP $ toClosedType k (normalizeTypeG tyG)
+  return (ty1 == (AlexPn 0 0 0 <$ ty2))
+
+prop_normalizeTypeSame :: TyProp
+prop_normalizeTypeSame k tyG tyQ = isSafe $ do
+  ty <- withExceptT GenErrorP tyQ
+  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
+  tyN1 <- withExceptT TypeErrorP $ case normalizeTypeAgda (AlexPn 0 0 0 <$ tyDB) of
+    Just tyN -> return tyN
+    Nothing  -> throwError undefined -- TODO
+  ty1 <- withExceptT FVErrorP $ unDeBruijnTy tyN1
```

Comment:
Why not merge most of the tests in this file into a single one? Here you do a lot of checks and transformations over and over again. You could stick everything into a single giant test and avoid regenerating and renormalizing types in different tests.

Reply:
This sounds like a good plan from an efficiency point of view and in terms of improving the testing code. But, wouldn't this make the tests less granular - i.e., if it fails it wouldn't be so clear which aspect was failing?

Reply:
Rrright... Given that we only generate a few thousands of well-kinded types, can we just materialize them into a list using [`search`](http://hackage.haskell.org/package/lazy-search-0.1.2.1/docs/Control-Search.html#v:search) and then manipulate that list directly and use the regular `Test.Tasty` stuff for creating test cases? Both `search` and `ctrex'` call `searchRaw'` under the hood, so this should cause us problems.